RecoParticleFlow/PFProducer/plugins/linkers/TrackAndHCALLinker.cc

```diff
@@ -45,89 +64,91 @@ class TrackAndHCALLinker : public BlockElementLinkerBase {
 
 DEFINE_EDM_PLUGIN(BlockElementLinkerFactory, TrackAndHCALLinker, "TrackAndHCALLinker");
 
-bool TrackAndHCALLinker::linkPrefilter(const reco::PFBlockElement* elem1, const reco::PFBlockElement* elem2) const {
+bool TrackAndHCALLinker::linkPrefilter(size_t ielem1,
+                                       size_t ielem2,
+                                       reco::PFBlockElement::Type type1,
+                                       reco::PFBlockElement::Type type2,
+                                       const PFTables& tables,
+                                       const reco::PFMultiLinksIndex& multilinks,
+                                       const reco::PFBlockElement* elem1,
+                                       const reco::PFBlockElement* elem2) const {
   bool result = false;
-  // Track-HCAL KDTree multilinks are stored to track's elem
-  switch (elem1->type()) {
+  switch (type1) {
     case reco::PFBlockElement::TRACK:
-      result = (elem1->isMultilinksValide(elem2->type()) && !elem1->getMultilinks(elem2->type()).empty() &&
-                elem2->isMultilinksValide(elem1->type()));
+      result = multilinks.isValid(ielem1, type1, type2);
```

Comment:
I need to think if this check is doing the same thing. In the "original" code, the check is done to make sure both elem1 and elem2 have multiliks of the "right" link type.

Reply:
Indeed, we pass type1 and type2 to check the multilinks of this element by type, you can verify in https://github.com/cms-sw/cmssw/pull/31191/files#diff-d8b43a58384a8e7415021b981d132c56R53 how it's implemented.

There are no reco differences, so I have no reason to believe that it behaves differently.

Reply:
Right. This is not about reco differences, but about the prefilter efficiency.
Previously, it was checking that elem1 has a linked element of the right type (but not necessarily elem2 which required static_cast in testlink), and also elem2 has a linked element of the right type (but not necessarily elem1 is the right type.)
I think isValid(ielem2,type2,type1) can be considered, after making sure multiliks are filled to make it work, but also
with your update of multinks, you could consider simply using isLinked(ielem1,ielem2,type1,type2)?
And, possibly you can store distance in multinks? which can simplify what you have to do testlink.